feat(Simple Vector Store Node): Implement store cleaning based on age/used memory

OlegIvaniv opened the pull request:
## Summary

This PR adds memory management capabilities to the in-memory vector stores used in the VectorStoreInMemory nodes:

- Adds configurable memory limits through the `N8N_VECTOR_STORE_MAX_MEMORY` environment variable, defaulting to unlimited (-1)
- Implements time-to-live (TTL) cleanup for inactive vector stores via `N8N_VECTOR_STORE_TTL_HOURS` environment variable, defaulting to no cleanup (-1)
- Provides memory usage tracking and estimation for vector stores
- Creates a cleanup service that removes vector stores based on memory pressure and inactivity
- Restructures the existing manager into a more modular design with separate responsibilities

## Technical Details

- Memory calculation is based on embedding dimensions, document content length, and metadata size
- Cleanup prioritizes inactive stores first, then removes oldest stores when memory limits are reached
- Default behavior maintains unlimited memory usage (-1) and no TTL cleanup (-1) for backward compatibility
- Memory usage statistics are available for debugging and monitoring
- Simple cache optimization for efficient cleanup of oldest stores

## Testing

- Added unit tests for the memory calculator, cleanup service, and manager components
- Tested with various memory limit configurations and document batch sizes
- Validated behavior during memory pressure scenarios
- Verified TTL cleanup works correctly with different time configurations

## Configuration

- Set `N8N_VECTOR_STORE_MAX_MEMORY=100` to limit vector store memory usage to 100MB
- Set `N8N_VECTOR_STORE_TTL_HOURS=24` to remove vector stores inactive for 24+ hours
- Both settings can be changed without restarting n8n (picked up on next workflow execution)

## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

OlegIvaniv committed:
feat(Simple Vector Store Node): Implement store cleaning based on age/used memory

OlegIvaniv committed:
Merge branch 'master' into ai-775-limit-the-memory-consumption-in-in-memory-vector-store-chat

OlegIvaniv committed:
After merge clean up

jeanpaul reviewed (changes requested):
Nice separation of concerns, thanks! (It's a little jarring to see a file of 52 lines being so blown up, but I understand the necessity :-))

I have a few small comments, but all up for discussion

jeanpaul commented:
> Another thing to consider... When you are working on a vector store with key `vector_store` only, and you reach your limit adding X vectors to it, it will clear out that vector store `vector_store` instead of removing X vectors and/or dropping the last few vector that don't fit in memory. Something that should definitely be addressed in a doc and maybe even a warning in the node output

Ah, sorry, it will apparently clear the current `vector_store` store, and then add the X vectors to it?

OlegIvaniv committed:
Update notice and remove additional safety buffer

OlegIvaniv commented on packages/@n8n/nodes-langchain/nodes/vector_store/shared/MemoryManager/MemoryVectorStoreManager.ts:
```diff
@@ -0,0 +1,311 @@
+	/**
+	 * Get statistics about the vector store memory usage
+	 */
+	getStats(): VectorStoreStats {
```
Nowhere right now, I originally used it for logging inside `populateVectorStore` but reduced the logging. But as you mentioned, it provides an easy way to tests the memory so I'd vote to keep it

OlegIvaniv commented:
@jeanpaul Thanks for review. I removed the additional buffer offset, added notice & created docs [PR](https://github.com/n8n-io/n8n-docs/pull/3002). Could you have another look, pls?

jeanpaul commented on packages/@n8n/nodes-langchain/nodes/vector_store/shared/MemoryManager/MemoryVectorStoreManager.ts:
```diff
@@ -0,0 +1,311 @@
+	/**
+	 * Set up timer for TTL-based cleanup
+	 */
+	private setupTtlCleanup(): void {
```
I disagree -- to me cleanup things seem to be the responsibility of the cleanup service, which would also take care of setting the interval. Doesn't have to be exactly this code in the constructor. But, I'm okay with leaving this if you feel strongly about it :-)

janober commented:
Got released with [`n8n@1.85.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.85.0)
